cronjobs/sftp_connection.py

```diff
@@ -0,0 +1,101 @@
+# Python 3 module
+# -*- coding: utf-8 -*-
+import paramiko
+import os
+import util
+
+class SFTPConnection:  
+  _host = ""
+  _username = ""
+  _password = ""
+  _port = 22
+  _sftp = None
+
+  def __init__(self, hostname, username, password, port=None):
+    self._host = hostname
+    self._username = username
+    self._password = password
+    self._sftp = paramiko.SSHClient()
+
+    if port is not None:
+      self._port = port
+    
+    self._connect()
+
+    
+  def getSFTP(self):
+    return self._sftp
+  
+
+  def _connect(self):
+    util.Info(f"Connecting to {self._host}")
+    try:
+      self._sftp.set_missing_host_key_policy(paramiko.AutoAddPolicy())
+      self._sftp.connect(hostname=self._host, port=self._port, username=self._username, password=self._password, look_for_keys=False, timeout=120)
+    except Exception as excp:
+      util.Error(f"Failed to connect to SFTP server! ({excp})")
+    else:
+      util.Info(f"Connected to server {self._host}")
+
+
+  def changeDirectory(self, remote_dir_path):
+    util.Info(f"change directory to: {remote_dir_path}")
+    sftp_client = self._sftp.open_sftp()
+    try:
+      sftp_client.chdir(remote_dir_path)
+    except Exception as excp:
+      util.Error(f"failed to change directory: {remote_dir_path}! {excp}")
+      
+
+  def listDirectory(self, remote_dir_path='.'):
+    sftp_client = self._sftp.open_sftp()
+    try:
+      return sftp_client.listdir(path=remote_dir_path)
+    except Exception as excp:
+      util.Error(f"failed to list directory ({excp})")
+
+
+  def downloadFile(self, remote_file_path, local_file_path=None):
+    if local_file_path is None:
+      local_file_path = remote_file_path
+
+    util.Info(f"downloading file from {remote_file_path} to {local_file_path}")
+    sftp_client = self._sftp.open_sftp()
```

Comment:
The sftp_client should be stored in a class variable, and open_sftp should be called only once in the __init__ function.